src/contract_client.rs

```diff
@@ -668,6 +677,193 @@ impl ModuleDeployHandle {
+    contract_name: &OwnedContractName,
+) -> Option<Type> {
+    // Remove the 'init_' prefix from the contract name.
+    let no_prefix_contract_name = &contract_name.to_string()[5..];
+    let contract_name_length = no_prefix_contract_name.len();
+
+    // Remove the 'contract_name.' prefix from the entrypoint name.
+    let no_prefix_receive_name = &receive_name.to_string()[contract_name_length + 1..];
+
+    schema
+        .get_receive_error_schema(no_prefix_contract_name, no_prefix_receive_name)
+        .ok()
+}
+
+/// Decode the smart contract logical revert reason and return a human-readable
+/// error string.
+///
+/// If the error is NOT caused by a smart contract logical revert, the
+/// `reject_reason` is already a human-readable error. No further decoding of
+/// the error is needed and as such this function returns `None`.
+/// An example of such a failure (rejected transaction) is the case when the
+/// transaction runs out of energy which is represented by the human-readable
+/// error variant "OutOfEnergy".
+///
+/// Step 1: If the error is caused by a smart contract logical revert coming
+/// from the `concordium-std` crate, this function decodes the error based on
+/// the error code definition in the `concordium-std` crate.
+///
+/// Step 2: If the error is caused by a smart contract logical revert coming
+/// from the smart contract itself, this function uses the provided
+/// `error_schema` and `return_value` to decode the `reject_reason` into a
+/// human-readable error string.
+///
+/// Disclaimer: A smart contract can have logic to overwrite/change the meaning
+/// of the error codes as defined in the `concordium-std` crate. While it is not
+/// advised to overwrite these error codes and is rather unusual to do so, this
+/// function decodes the error codes based on the
+/// definitions in the `concordium-std` crate (assuming they have not been
+/// overwritten with other meanings in the smart contract logic). No guarantee
+/// are given as such that the meaning of the decoded reject reason haven't been
+/// altered by the smart contract logic.
+pub fn decode_smart_contract_revert(
+    return_value: Option<ReturnValue>,
+    reject_reason: RejectReason,
+    schema: &Option<VersionedModuleSchema>,
+    receive_name: OwnedReceiveName,
+    contract_name: &OwnedContractName,
+) -> Result<Option<String>, RejectReasonDecodeError> {
+    match reject_reason {
+        RejectReason::RejectedReceive {
+            reject_reason: reject_reason_code,
+            contract_address: _,
+            receive_name: _,
+            parameter: _,
+        } => {
+            // Step 1: Try to decode the `reject_reason` using the `concordium-std`
+            // error codes.
+            if let Some(decoded_error) = decode_concordium_std_error(reject_reason_code) {
+                return Ok(Some(decoded_error));
+            }
+
+            // Step 2: Try to decode the `reject_reason` using the `error_schema` and the
```

Comment:
**Question (non-blocking):** Would it make sense to do step 2 first? Meaning, the schema defined errors take precedence to the standardized ones.

Reply:
I don't feel strongly about the order since the error codes would not overlap in general (it is rather unusual for a smart contract developer to overwrite error codes in the smart contract manually - I can not come up with a meaningful reason for doing it at all). But I changed the order now so that the `smart contract error` take precedence over the `concordium-std` error.

Reply:
okay, then let's stick to this order, the first step is also a lot cheaper.